### PR TITLE
Replace sui modal with a custom one

### PR DESCRIPTION
**Pull request by MiZiet**
If applied these changes will:
- replace sui modal with a custom one,
- move close modal button to the modal header,
- add modal close on esc button and on background click,
- add background to the modal action part.

### CODE REVIEW

File: packages/react-components/src/Modal/Base.tsx [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+import { Header } from './Header';
+import { Body, Overlay } from './styled';
+
+function Base (props: ModalProps): React.ReactElement<ModalProps> {
+  const { theme } = useContext<ThemeDef>(ThemeContext);
+  const { children, className = '', header, onClose, open = true, size = 'medium', testId = 'modal' } = props;
+
+  const listenKeyboard = useCallback((event: KeyboardEvent) => {
+    if (event.key === 'Escape' || event.keyCode === 27) {
```
Comment (wachulski) [inline packages/react-components/src/Modal/Base.tsx:19]: a const somewhere? e.g. ESC_KEYCODE?

**Review by wachulski — APPROVED** [review]
LGTM

**Comment by krzysztof-jelski** [comment]
I noticed some minor UI issues:
* the corners on the bottom (actions) part are not rounded, as opposed to those on top
image
* the placement of the whole modal relative to the page is a little bit different than before. Not sure it's a problem, though.

**Comment by krzysztof-jelski** [comment]
Also, not entirely sure about the "close on background click" feature. I mean I'm afraid about accidental clicks in case some data is already entered into the modal. Or for example when user has copied the mnemonic phrase, went to other app to paste it, came back and clicked outside the modal by accident. Still, I guess I'm inclined to try it out as this PR implements it and see how that works out.

**Review by krzysztof-jelski — CHANGES_REQUESTED** [review]
Bottom corners should be rounded.

File: packages/page-addresses/src/modals/Create.tsx [hunk 2]
```diff
@@ -114,7 +114,8 @@ function Create ({ onClose, onStatusChange }: Props): React.ReactElement<Props>
   );
 
   return (
-    <Modal header={t<string>('Add an address')}>
+    <Modal header={t<string>('Add an address')}
+      onClose={onClose}>
```
Comment (jacogr) [inline packages/page-addresses/src/modals/Create.tsx:118]:
```suggestion
    <Modal
      header={t<string>('Add an address')}
      onClose={onClose}
    >
```

File: packages/page-contracts/src/Codes/Add.tsx [hunk 3]
```diff
@@ -39,7 +39,8 @@ function Add ({ onClose }: Props): React.ReactElement {
   const isValid = isCodeHashValid && isNameValid && isAbiSupplied && isAbiValid;
 
   return (
-    <Modal header={t('Add an existing code hash')}>
+    <Modal header={t('Add an existing code hash')}
+      onClose={onClose}>
```
Comment (jacogr) [inline packages/page-contracts/src/Codes/Add.tsx:43]:
```suggestion
    <Modal
      header={t('Add an existing code hash')}
      onClose={onClose}
    >
```

File: packages/page-contracts/src/Codes/Upload.tsx [hunk 4]
```diff
@@ -131,7 +131,8 @@ function Upload ({ onClose }: Props): React.ReactElement {
   const invalidAbi = isAbiError || !isAbiSupplied;
 
   return (
-    <Modal header={t('Upload & deploy code {{info}}', { replace: { info: `${step}/2` } })}>
+    <Modal header={t('Upload & deploy code {{info}}', { replace: { info: `${step}/2` } })}
+      onClose={onClose}>
```
Comment (jacogr) [inline packages/page-contracts/src/Codes/Upload.tsx:135]:
```suggestion
    <Modal
      header={t('Upload & deploy code {{info}}', { replace: { info: `${step}/2` } })}
      onClose={onClose}
    >
```

File: packages/page-contracts/src/Contracts/Add.tsx [hunk 5]
```diff
@@ -64,7 +64,8 @@ function Add ({ onClose }: Props): React.ReactElement {
   const isValid = isAddressValid && isNameValid && isAbiValid;
 
   return (
-    <Modal header={t('Add an existing contract')}>
+    <Modal header={t('Add an existing contract')}
+      onClose={onClose}>
```
Comment (jacogr) [inline packages/page-contracts/src/Contracts/Add.tsx:68]:
```suggestion
    <Modal
      header={t('Add an existing contract')}
      onClose={onClose}
    >
```

... eslint should really sort these out and format correctly.

File: packages/page-contracts/src/Contracts/Deploy.tsx [hunk 6]
```diff
@@ -117,7 +117,8 @@ function Deploy ({ codeHash, constructorIndex = 0, onClose, setConstructorIndex
   const isValid = isNameValid && isEndowmentValid && weight.isValid && isAccountIdValid && isSaltValid;
 
   return (
-    <Modal header={t('Deploy a contract')}>
+    <Modal header={t('Deploy a contract')}
+      onClose={onClose}>
```
Comment (jacogr) [inline packages/page-contracts/src/Contracts/Deploy.tsx:121]:
```suggestion
    <Modal
      header={t('Deploy a contract')}
      onClose={onClose}
    >
```

File: packages/page-society/src/Candidates/CandidateVoting.tsx [hunk 7]
```diff
@@ -29,7 +29,8 @@ function CandidateVoting ({ candidateId, isMember, ownMembers }: Props): React.R
   return (
     <>
       {isVisible && (
-        <Modal header={t<string>('Vote for candidate')}>
+        <Modal header={t<string>('Vote for candidate')}
+          onClose={toggleVisible}>
```
Comment (jacogr) [inline packages/page-society/src/Candidates/CandidateVoting.tsx:33]:
```suggestion
        <Modal
          header={t<string>('Vote for candidate')}
          onClose={toggleVisible}
        >
```

File: packages/page-society/src/Overview/DefenderVoting.tsx [hunk 8]
```diff
@@ -28,7 +28,8 @@ function DefenderVoting ({ isMember, ownMembers }: Props): React.ReactElement<Pr
   return (
     <>
       {isVisible && (
-        <Modal header={t<string>('Vote for defender')}>
+        <Modal header={t<string>('Vote for defender')}
+          onClose={toggleVisible}>
```
Comment (jacogr) [inline packages/page-society/src/Overview/DefenderVoting.tsx:32]:
```suggestion
        <Modal
          header={t<string>('Vote for defender')}
          onClose={toggleVisible}
        >
```

File: packages/react-components/src/Modal/Actions.tsx [hunk 9]
```diff
@@ -4,23 +4,22 @@
...
-    </SUIModal.Actions>
-  );
-}
+const Actions: React.FC<ActionsProps> = ({ children, className = '' }) => (
+  <div className={className}>
+    <Button.Group>
+      {children}
+    </Button.Group>
+  </div>
+);
```
Comment (jacogr) [inline packages/react-components/src/Modal/Actions.tsx:17]:
```suggestion
function Actions ({ children, className = '' }: ActionsProps) {
  return (
    <div className={className}>
      <Button.Group>
        {children}
      </Button.Group>
    </div>
  );
}
```

File: packages/react-components/src/Modal/Header.tsx [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+    {header && (
+      <h1>{header}</h1>
+    )}
+    <Button
+      dataTestId='close-modal'
+      icon='times'
+      onClick={onClose}
+    />
+  </ModalHeader>
+);
```
Comment (jacogr) [inline packages/react-components/src/Modal/Header.tsx:22]:
```suggestion
function Header ({ header, onClose }: HeaderProps) {
  return (
    <ModalHeader className='header'>
      {header && (
        <h1>{header}</h1>
      )}
      <Button
        dataTestId='close-modal'
        icon='times'
        onClick={onClose}
      />
    </ModalHeader>
  );
}
```

File: packages/react-components/src/Modal/Header.tsx [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+
+const ModalHeader = styled.div`
```
Comment (jacogr) [inline packages/react-components/src/Modal/Header.tsx:24]: Although this is the way styled-components suggest, we really don't use it anywhere. It makes it incredibly difficult, imho, to reason about, hence it not being used.

Rather just `<div className='header'>` and what the export in React.memo(Header) with the styles. (As done elsewhere)

File: packages/react-components/src/Modal/styled.tsx [hunk 12]
```diff
@@ -0,0 +1,45 @@
+// Copyright 2017-2021 @polkadot/react-components authors & contributors
+// SPDX-License-Identifier: Apache-2.0
+
+import type { BodyProps, ContentProps } from './types';
+
+import styled from 'styled-components';
+
+export const Overlay = styled.div`
```
Comment (jacogr) [inline packages/react-components/src/Modal/styled.tsx:8]: Don't like these types of component, as explained above. It really, really makes refactoring an absolute nightmare. As much as styled-component recommends exactly this approach, it is not something that gets used here.

File: packages/react-components/src/styles/semantic.ts [hunk 13]
```diff
@@ -123,50 +123,6 @@ export default (_theme: ThemeDef): string => `
     z-index: 1;
   }
 
-  .ui.modal {
```
Comment (jacogr) [inline packages/react-components/src/styles/semantic.ts:126]: +1 - generally would have been forgotten about, great cleanup.

File: packages/react-components/src/Modal/Base.tsx (hunk 1 above)
Comment (MiZiet) [reply packages/react-components/src/Modal/Base.tsx]: ✅

File: packages/page-addresses/src/modals/Create.tsx (hunk 2 above)
Comment (MiZiet) [reply packages/page-addresses/src/modals/Create.tsx]: ✅

File: packages/page-contracts/src/Codes/Add.tsx (hunk 3 above)
Comment (MiZiet) [reply packages/page-contracts/src/Codes/Add.tsx]: ✅

File: packages/page-contracts/src/Codes/Upload.tsx (hunk 4 above)
Comment (MiZiet) [reply packages/page-contracts/src/Codes/Upload.tsx]: ✅

File: packages/page-contracts/src/Contracts/Add.tsx (hunk 5 above)
Comment (MiZiet) [reply packages/page-contracts/src/Contracts/Add.tsx]: ✅

File: packages/page-contracts/src/Contracts/Deploy.tsx (hunk 6 above)
Comment (MiZiet) [reply packages/page-contracts/src/Contracts/Deploy.tsx]: ✅

File: packages/page-society/src/Overview/DefenderVoting.tsx (hunk 8 above)
Comment (MiZiet) [reply packages/page-society/src/Overview/DefenderVoting.tsx]: ✅

File: packages/react-components/src/Modal/Actions.tsx (hunk 9 above)
Comment (MiZiet) [reply packages/react-components/src/Modal/Actions.tsx]: ✅

File: packages/react-components/src/Modal/Header.tsx (hunk 10 above)
Comment (MiZiet) [reply packages/react-components/src/Modal/Header.tsx]: ✅

File: packages/react-components/src/Modal/Header.tsx (hunk 11 above)
Comment (MiZiet) [reply packages/react-components/src/Modal/Header.tsx]: ✅

File: packages/react-components/src/Modal/styled.tsx (hunk 12 above)
Comment (MiZiet) [reply packages/react-components/src/Modal/styled.tsx]: ✅

**Comment by MiZiet** [comment]
> Bottom corners should be rounded.

Done ✅

**Review by krzysztof-jelski — APPROVED** [review]

**Review by jacogr — APPROVED** [review]
Not 100% convinced on -

- the close button at the top-right (but I'm guessing this may be a muscle memory issue on my side)
- the background for the buttons, looks slightly "off", but don't have better ideas, there is probably something to be said for making it stand out.

Apart from that seems to do what it says on the tin, thank you.

**Comment by MiZiet** [comment]
Great, we will try to think and improve the buttons section but preferably in a separate pr. I'd like to merge it but I don't have merging permission. Could you merge it @jacogr ?

**Pull request merged**